libkrun/libkruntime.c

```diff
@@ -37,86 +79,254 @@ clock_gettime_monotonic()
+{
+    int core;
+
+    for (core = 0; core < num_msr_nodes; core++) {
+        config_fixed_ctr1(core, enable);
+        write_msr(core, PCTR_IA32_PERF_FIXED_CTR1, 0); // reset
+    }
+}
+
+void
+close_fd(int fd)
+{
+    int ok = close(fd);
+    if (ok < 0) {
+        perror("close");
+        exit(EXIT_FAILURE);
+    }
+}
+
+int
+get_fixed_pctr1_width()
 {
-    u_int32_t hi, lo;
+    uint32_t eax, edx;
+    int fixed_ctr_width, arch_ctr_vers, num_fixed_ctrs;
 
-    __asm volatile (
-        "cpuid\n\t"             // Barrier (trashes e[abcd]x)
-        "rdtsc\n\t"             // Put TSR in edx:eax, also trashes ecx
-        "mov %%edx, %0\n\t"     // Copy out the high 32-bits of TSR
-        "mov %%eax, %1\n\t"     // Copy out the low 32-bits of TSR
-        // ---
-        : "=r" (hi), "=r" (lo)              // outputs
-        :                                   // inputs
-        : "%rax", "%rbx", "%rcx", "%rdx");  // other clobbers
-    return ((u_int64_t) hi << 32) | (u_int64_t) lo;
+    asm volatile(
+        "mov $0xa, %%eax\n\t"       // pctr leaf
+        "cpuid\n\t"
+        : "=a" (eax), "=d" (edx)    // out
+        :                           // in
+        : "ebx", "ecx");            // clobber
+
+    /* eax
+     *   0-4:  number of fixed-func counters
+     *   5-12: width of counters
+     */
+    num_fixed_ctrs = edx & 0x1f;
+    fixed_ctr_width = (edx & 0x1fe0) >> 5;
+
+    /* eax
+     *   0-7:  architectural counter version
+     *   8-31: reserved
+     */
+    arch_ctr_vers = eax & (0xff);
+
+    // fixed function perf ctrs appeared on arch counter version 2
+    if (arch_ctr_vers < 2) {
+        printf("arch pctr version >=2 is required! got %d\n", arch_ctr_vers);
+        exit(EXIT_FAILURE);
+    }
+
+    // We are interested in IA32_FIXED_CTR1, (i.e. the second fixed counter)
+    if (num_fixed_ctrs < 2) {
+        fprintf(stderr, "too few fixed-function counters: %d\n",
+            num_fixed_ctrs);
+        exit(EXIT_FAILURE);
+    }
+
+    return fixed_ctr_width;
+}
+#endif // linux && !TRAVIS
+
+void
+libkruntime_init(void)
+{
+#if defined(__linux__) && !defined(TRAVIS)
+    int core;
+
+    /* See how wide the counter values are and make an appropriate mask */
+    pctr_val_mask = ((uint64_t) 1 << get_fixed_pctr1_width()) - 1;
+
+    /* Set up MSR device node handles */
+    num_msr_nodes = sysconf(_SC_NPROCESSORS_ONLN);
+
+    msr_r_nodes = calloc(num_msr_nodes, sizeof(int));
+    if (msr_r_nodes == NULL) {
+        perror("calloc");
+        exit(EXIT_FAILURE);
+    }
+
+    msr_w_nodes = calloc(num_msr_nodes, sizeof(int));
+    if (msr_w_nodes == NULL) {
+        perror("calloc");
+        exit(EXIT_FAILURE);
+    }
+
+    for (core = 0; core < num_msr_nodes; core++) {
+        msr_r_nodes[core] = open_msr_node(core, O_RDONLY);
+        msr_w_nodes[core] = open_msr_node(core, O_WRONLY);
+    }
+
+    /* Configure and reset CPU_CLK_UNHALTED.CORE on all CPUs */
+    config_fixed_ctr1_all_cores(1);
+#else
+    /*
```

Comment:
At the moment, if you compile on an unsupported platform, things will silently compile, but without telling you that you're missing support for a feature that your platform may support. It's much easier, IMHO, to have the code raise an error which says "look at this bit!" rather than having to grep it later (assuming you even know what to grep for).